File: src/main/java/com/dh/ondot/member/domain/service/MemberService.java
package com.dh.ondot.member.domain.service;

import com.dh.ondot.member.application.command.OnboardingCommand;
import com.dh.ondot.member.core.exception.AlreadyOnboardedException;
import com.dh.ondot.member.core.exception.NotFoundMemberException;
import com.dh.ondot.member.domain.Member;
import com.dh.ondot.member.domain.dto.UserInfo;
import com.dh.ondot.member.domain.enums.OauthProvider;
import com.dh.ondot.member.domain.repository.MemberRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class MemberService {
    private final MemberRepository memberRepository;

    public Member getMemberIfExists(Long memberId) {
        return memberRepository.findById(memberId)
                .orElseThrow(() -> new NotFoundMemberException(memberId));
    }

    public Member getAndValidateAlreadyOnboarded(Long memberId) {
        Member member = getMemberIfExists(memberId);
        if (!member.isNewMember()) {
            throw new AlreadyOnboardedException(member.getId());
        }
        return member;
    }

    public Long getTotalMemberCount() {
        return memberRepository.count();
    }

    @Transactional
    public Member findOrRegisterOauthMember(UserInfo userInfo, OauthProvider oauthProvider) {
        return memberRepository.findByOauthInfo(userInfo.oauthProviderId(), oauthProvider)
                .orElseGet(() -> registerMemberWithOauth(userInfo, oauthProvider));
    }

    private Member registerMemberWithOauth(UserInfo userInfo, OauthProvider oauthProvider) {
        Member newMember = Member.registerWithOauth(
                userInfo.email(),
                oauthProvider,
                userInfo.oauthProviderId()
        );
        return memberRepository.save(newMember);
    }

    @Transactional
    public Member updatePreparationTime(Long memberId, Integer preparationTime) {
        Member member = getMemberIfExists(memberId);
        member.updatePreparationTime(preparationTime);
        return member;
    }

    @Transactional
    public Member updateOnboardingInfo(Member member, OnboardingCommand command) {
        member.updateOnboarding(
                command.preparationTime(), command.alarmMode(),
                command.isSnoozeEnabled(), command.snoozeInterval(), command.snoozeCount(),
                command.soundCategory(), command.ringTone(), command.volume()
        );
        return member;
    }

    @Transactional
    public void deleteMember(Long memberId) {
        memberRepository.deleteById(memberId);
    }
}

File: src/main/kotlin/com/dh/ondot/member/domain/service/MemberService.kt
package com.dh.ondot.member.domain.service

import com.dh.ondot.member.application.command.OnboardingCommand
import com.dh.ondot.member.core.exception.AlreadyOnboardedException
import com.dh.ondot.member.core.exception.NotFoundMemberException
import com.dh.ondot.member.domain.Member
import com.dh.ondot.member.domain.dto.UserInfo
import com.dh.ondot.member.domain.enums.OauthProvider
import com.dh.ondot.member.domain.repository.MemberRepository
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional(readOnly = true)
class MemberService(
    private val memberRepository: MemberRepository,
) {
    fun getMemberIfExists(memberId: Long): Member =
        memberRepository.findById(memberId)
            .orElseThrow { NotFoundMemberException(memberId) }

    fun getAndValidateAlreadyOnboarded(memberId: Long): Member {
        val member = getMemberIfExists(memberId)
        if (!member.isNewMember()) {
            throw AlreadyOnboardedException(member.id)
        }
        return member
    }

    fun getTotalMemberCount(): Long = memberRepository.count()

    @Transactional
    fun findOrRegisterOauthMember(userInfo: UserInfo, oauthProvider: OauthProvider): Member =
        memberRepository.findByOauthInfo(userInfo.oauthProviderId, oauthProvider)
            ?: registerMemberWithOauth(userInfo, oauthProvider)

    private fun registerMemberWithOauth(userInfo: UserInfo, oauthProvider: OauthProvider): Member {
        val newMember = Member.registerWithOauth(
            userInfo.email.orEmpty(),
            oauthProvider,
            userInfo.oauthProviderId,
        )
        return memberRepository.save(newMember)
    }

    @Transactional
    fun updatePreparationTime(memberId: Long, preparationTime: Int): Member {
        val member = getMemberIfExists(memberId)
        member.updatePreparationTime(preparationTime)
        return member
    }

    @Transactional
    fun updateOnboardingInfo(member: Member, command: OnboardingCommand): Member {
        member.updateOnboarding(
            command.preparationTime, command.alarmMode,
            command.isSnoozeEnabled, command.snoozeInterval, command.snoozeCount,
            command.soundCategory, command.ringTone, command.volume,
        )
        return member
    }

    @Transactional
    fun deleteMember(memberId: Long) {
        memberRepository.deleteById(memberId)
    }

    @Transactional
    fun updateDailyReminderEnabled(memberId: Long, enabled: Boolean): Member {
        val member = getMemberIfExists(memberId)
        member.updateDailyReminderEnabled(enabled)
        return member
    }

    fun findAllDailyReminderEnabledMembers(): List<Member> =
        memberRepository.findAllByDailyReminderEnabledTrue()
}
